### PR TITLE
sbang: add support for php

### LINKED ISSUE #15543: Installation issue: php
### Spack version
```console
$ spack --version
0.14.0-245-847d5d3
```

### Steps to reproduce the issue
NB: I had to apply the fixes in #15435 and updated the version to 7.4.4 to get the build to work without errors on sqlite3.

I followed @adamjstewart's suggestion in #15435 to run the install with `-d` so the following output is long, though limited to the failing steps. It's certainly an issue with the shebang
fixup, but I can't track down which file(s) are causing the issue. If anyone can give me hints, happy to fix/track down further

```console
$ spack -d install php
...
==> [2020-03-18-14:08:46.269623, 23586] Warning: Module file already exists : skipping creation
file : /storage/epp2/epp_software/spack-devel/share/spack/modules/linux-centos7-skylake/php-7.4.4-gcc-9.3.0-wtyw6jf
spec : php@7.4.4%gcc@9.3.0 arch=linux-centos7-skylake ^autoconf@2.69%gcc@9.3.0 arch=linux-centos7-skylake ^automake@1.16.1%gcc@9.3.0 arch=linux-centos7-skylake ^bison@3.4.2%gcc@9.3.0 patches=89aa362716d898edd0b5c6ae4208dc1b6992887774848a09e8021afd676f7d61 arch=linux-centos7-skylake ^bzip2@1.0.8%gcc@9.3.0+shared arch=linux-centos7-skylake ^diffutils@3.7%gcc@9.3.0 arch=linux-centos7-skylake ^findutils@4.6.0%gcc@9.3.0 patches=84b916c0bf8c51b7e7b28417692f0ad3e7030d1f3c248ba77c42ede5c1c5d11e,bd9e4e5cc280f9753ae14956c4e4aa17fe7a210f55dd6c84aa60b12d106d47a2 arch=linux-centos7-skylake ^flex@2.6.4%gcc@9.3.0+lex patches=09c22e5c6fef327d3e48eb23f0d610dcd3a35ab9207f12e0f875701c677978d3 arch=linux-centos7-skylake ^gettext@0.20.1%gcc@9.3.0+bzip2+curses+git~libunistring+libxml2+tar+xz arch=linux-centos7-skylake ^help2man@1.47.11%gcc@9.3.0 arch=linux-centos7-skylake ^libiconv@1.16%gcc@9.3.0 arch=linux-centos7-skylake ^libsigsegv@2.12%gcc@9.3.0 arch=linux-centos7-skylake ^libtool@2.4.6%gcc@9.3.0 arch=linux-centos7-skylake ^libxml2@2.9.9%gcc@9.3.0~python arch=linux-centos7-skylake ^m4@1.4.18%gcc@9.3.0 patches=3877ab548f88597ab2327a2230ee048d2d07ace1062efe81fc92e91b7f39cd00,fc9b61654a3ba1a8d6cd78ce087e7c96366c290bc8d2c299f09828d793b853c8 +sigsegv arch=linux-centos7-skylake ^ncurses@6.2%gcc@9.3.0~symlinks+termlib arch=linux-centos7-skylake ^perl@5.16.3%gcc@9.3.0+cpanm patches=0eac10ed90aeb0459ad8851f88081d439a4e41978e586ec743069e8b059370ac +shared+threads arch=linux-centos7-skylake ^pkgconf@1.6.3%gcc@9.3.0 arch=linux-centos7-skylake ^re2c@1.2.1%gcc@9.3.0 arch=linux-centos7-skylake ^readline@8.0%gcc@9.3.0 arch=linux-centos7-skylake ^sqlite@3.30.1%gcc@9.3.0~column_metadata+fts~functions~rtree arch=linux-centos7-skylake ^tar@1.32%gcc@9.3.0 arch=linux-centos7-skylake ^texinfo@6.5%gcc@9.3.0 patches=12f6edb0c6b270b8c8dba2ce17998c580db01182d871ee32b7b6e4129bd1d23a,1732115f651cff98989cb0215d8f64da5e0f7911ebf0c13b064920f088f2ffe1 arch=linux-centos7-skylake ^xz@5.2.4%gcc@9.3.0 arch=linux-centos7-skylake ^zlib@1.2.11%gcc@9.3.0+optimize+pic+shared arch=linux-centos7-skylake
==> [2020-03-18-14:08:46.605448, 23586] WRITE LOCK (spack-stage-php-7.4.4-wtyw6jfmd4h7jwbgg6ujc6asmxlf2roa): /tmp/username/spack-stage/.lock[6107926064061561343:1] [Releasing] 
==> [2020-03-18-14:08:46.605571, 23586] WRITE LOCK (spack-stage-php-7.4.4-wtyw6jfmd4h7jwbgg6ujc6asmxlf2roa): /tmp/username/spack-stage/.lock[6107926064061561343:1] [Released at 14:08:46.605551] 
==> [2020-03-18-14:08:46.616149, 23523] Error: UnicodeDecodeError: 'utf-8' codec can't decode byte 0xfc in position 1462: invalid start byte

/storage/epp2/epp_software/spack-devel/lib/spack/spack/build_environment.py:818, in child_process:
        815            tb_string = traceback.format_exc()
        816
        817            # build up some context from the offending package so we can
  >>    818            # show that, too.
        819            package_context = get_package_context(tb)
        820
        821            build_log = None

See build log for details:
  /tmp/username/spack-stage/spack-stage-php-7.4.4-wtyw6jfmd4h7jwbgg6ujc6asmxlf2roa/spack-build-out.txt

Traceback (most recent call last):
  File "/storage/epp2/epp_software/spack-devel/lib/spack/spack/build_environment.py", line 801, in child_process
    return_value = function()
  File "/storage/epp2/epp_software/spack-devel/lib/spack/spack/installer.py", line 1120, in build_process
    spack.hooks.post_install(pkg.spec)
  File "/storage/epp2/epp_software/spack-devel/lib/spack/spack/hooks/__init__.py", line 59, in __call__
    hook(*args, **kwargs)
  File "/storage/epp2/epp_software/spack-devel/lib/spack/spack/hooks/sbang.py", line 115, in post_install
    filter_shebangs_in_directory(directory, filenames)
  File "/storage/epp2/epp_software/spack-devel/lib/spack/spack/hooks/sbang.py", line 102, in filter_shebangs_in_directory
    filter_shebang(path)
  File "/storage/epp2/epp_software/spack-devel/lib/spack/spack/hooks/sbang.py", line 40, in filter_shebang
    original = original.decode(encoding='UTF-8')
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xfc in position 1462: invalid start byte
==> [2020-03-18-14:08:46.616385, 23523] Error: Failed to install php due to ChildError: UnicodeDecodeError: 'utf-8' codec can't decode byte 0xfc in position 1462: invalid start byte
/storage/epp2/epp_software/spack-devel/lib/spack/spack/build_environment.py:818, in child_process:
        815            tb_string = traceback.format_exc()
        816
        817            # build up some context from the offending package so we can
  >>    818            # show that, too.
        819            package_context = get_package_context(tb)
        820
        821            build_log = None

See build log for details:
  /tmp/username/spack-stage/spack-stage-php-7.4.4-wtyw6jfmd4h7jwbgg6ujc6asmxlf2roa/spack-build-out.txt
Traceback (most recent call last):
  File "/storage/epp2/epp_software/spack-devel/lib/spack/spack/build_environment.py", line 801, in child_process
    return_value = function()
  File "/storage/epp2/epp_software/spack-devel/lib/spack/spack/installer.py", line 1120, in build_process
    spack.hooks.post_install(pkg.spec)
  File "/storage/epp2/epp_software/spack-devel/lib/spack/spack/hooks/__init__.py", line 59, in __call__
    hook(*args, **kwargs)
  File "/storage/epp2/epp_software/spack-devel/lib/spack/spack/hooks/sbang.py", line 115, in post_install
    filter_shebangs_in_directory(directory, filenames)
  File "/storage/epp2/epp_software/spack-devel/lib/spack/spack/hooks/sbang.py", line 102, in filter_shebangs_in_directory
    filter_shebang(path)
  File "/storage/epp2/epp_software/spack-devel/lib/spack/spack/hooks/sbang.py", line 40, in filter_shebang
    original = original.decode(encoding='UTF-8')
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xfc in position 1462: invalid start byte

==> [2020-03-18-14:08:46.616495, 23523] Flagging php as failed: UnicodeDecodeError: 'utf-8' codec can't decode byte 0xfc in position 1462: invalid start byte
/storage/epp2/epp_software/spack-devel/lib/spack/spack/build_environment.py:818, in child_process:
        815            tb_string = traceback.format_exc()
        816
        817            # build up some context from the offending package so we can
  >>    818            # show that, too.
        819            package_context = get_package_context(tb)
        820
        821            build_log = None

See build log for details:
  /tmp/username/spack-stage/spack-stage-php-7.4.4-wtyw6jfmd4h7jwbgg6ujc6asmxlf2roa/spack-build-out.txt
Traceback (most recent call last):
  File "/storage/epp2/epp_software/spack-devel/lib/spack/spack/build_environment.py", line 801, in child_process
    return_value = function()
  File "/storage/epp2/epp_software/spack-devel/lib/spack/spack/installer.py", line 1120, in build_process
    spack.hooks.post_install(pkg.spec)
  File "/storage/epp2/epp_software/spack-devel/lib/spack/spack/hooks/__init__.py", line 59, in __call__
    hook(*args, **kwargs)
  File "/storage/epp2/epp_software/spack-devel/lib/spack/spack/hooks/sbang.py", line 115, in post_install
    filter_shebangs_in_directory(directory, filenames)
  File "/storage/epp2/epp_software/spack-devel/lib/spack/spack/hooks/sbang.py", line 102, in filter_shebangs_in_directory
    filter_shebang(path)
  File "/storage/epp2/epp_software/spack-devel/lib/spack/spack/hooks/sbang.py", line 40, in filter_shebang
    original = original.decode(encoding='UTF-8')
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xfc in position 1462: invalid start byte

==> [2020-03-18-14:08:46.619790, 23523] WRITE LOCK (php): /storage/epp2/epp_software/spack-devel/opt/spack/.spack-db/prefix_failures[6519162299755759610:1] [Acquiring] 
==> [2020-03-18-14:08:46.621240, 23523] write locking [6519162299755759610:1]: timeout None sec
==> [2020-03-18-14:08:46.621833, 23523] WRITE LOCK (php): /storage/epp2/epp_software/spack-devel/opt/spack/.spack-db/prefix_failures[6519162299755759610:1] [Acquired at 14:08:46.621807] 
==> [2020-03-18-14:08:46.621940, 23523] READ LOCK (database): /storage/epp2/epp_software/spack-devel/opt/spack/.spack-db/lock[0:0] [Acquiring] 
==> [2020-03-18-14:08:46.623453, 23523] read locking [0:0]: timeout 3 sec
==> [2020-03-18-14:08:46.623837, 23523] READ LOCK (database): /storage/epp2/epp_software/spack-devel/opt/spack/.spack-db/lock[0:0] [Acquired at 14:08:46.623818] 
==> [2020-03-18-14:08:46.632693, 23523] READ LOCK (database): /storage/epp2/epp_software/spack-devel/opt/spack/.spack-db/lock[0:0] [Releasing] 
==> [2020-03-18-14:08:46.633711, 23523] READ LOCK (database): /storage/epp2/epp_software/spack-devel/opt/spack/.spack-db/lock[0:0] [Released at 14:08:46.633688] 
==> [2020-03-18-14:08:48.101447, 23523] ChildError: UnicodeDecodeError: 'utf-8' codec can't decode byte 0xfc in position 1462: invalid start byte
/storage/epp2/epp_software/spack-devel/lib/spack/spack/build_environment.py:818, in child_process:
        815            tb_string = traceback.format_exc()
        816
        817            # build up some context from the offending package so we can
  >>    818            # show that, too.
        819            package_context = get_package_context(tb)
        820
        821            build_log = None

See build log for details:
  /tmp/username/spack-stage/spack-stage-php-7.4.4-wtyw6jfmd4h7jwbgg6ujc6asmxlf2roa/spack-build-out.txt
Traceback (most recent call last):
  File "/storage/epp2/epp_software/spack-devel/lib/spack/spack/build_environment.py", line 801, in child_process
    return_value = function()
  File "/storage/epp2/epp_software/spack-devel/lib/spack/spack/installer.py", line 1120, in build_process
    spack.hooks.post_install(pkg.spec)
  File "/storage/epp2/epp_software/spack-devel/lib/spack/spack/hooks/__init__.py", line 59, in __call__
    hook(*args, **kwargs)
  File "/storage/epp2/epp_software/spack-devel/lib/spack/spack/hooks/sbang.py", line 115, in post_install
    filter_shebangs_in_directory(directory, filenames)
  File "/storage/epp2/epp_software/spack-devel/lib/spack/spack/hooks/sbang.py", line 102, in filter_shebangs_in_directory
    filter_shebang(path)
  File "/storage/epp2/epp_software/spack-devel/lib/spack/spack/hooks/sbang.py", line 40, in filter_shebang
    original = original.decode(encoding='UTF-8')
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xfc in position 1462: invalid start byte

```

### Platform and user environment
```commandline
$ uname -a 
Linux hostname 3.10.0-1062.12.1.el7.x86_64 #1 SMP Tue Feb 4 23:02:59 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
$ lsb_release -d
Description:	CentOS Linux release 7.7.1908 (Core)
``` 
- `compilers.yaml`
  - Only change is that gcc@9.4.0 is installed by spack and then added to `compilers.yaml`, so `php` and deps are building with this compiler.

### General information

- [x] I have run `spack --version` and reported the version of Spack
- [x] I have searched the issues of this repo and believe this is not a duplicate

### PR DESCRIPTION
**Pull request by tkameyama**
fixed #15543

php script is processed only inside from `<?php` to `?>`.
another part is output to stdout.
The PR change `<?php #!` to sbang line.

php package installed phar script written by php.
But phar include binary, and phar embeded own sh1 checksum.
This PR insert sbang when generate phar script if it is needed.

### CODE REVIEW

**Comment by adamjstewart** [comment]
Looks like @tgamblin @muffgaga @luigi-calori @healther @michaelkuhn have modified our sbang support in the past. Could some of you review this PR?

**Comment by tkameyama** [comment]
ping  @tgamblin @muffgaga @luigi-calori @healther @michaelkuhn @adamjstewart

**Review by adamjstewart — APPROVED** [review]

**Comment by tgamblin** [comment]
@tkameyama: I modified this slightly to:

1. use a single-line PHP command for the second shebang (instead of spreading it over two lines)
2. to bring it up to date with #11598 

So, an `sbang`'d php script would look like this now:

```php
#!/bin/bash /Users/gamblin2/src/spack/bin/sbang
<?php #!/usr/bin/php ?>
<?php echo "hello world!\n" ?>
```

Can you review my changes?  If this is ok with you we can merge it.

File: bin/sbang [hunk 1]
```diff
@@ -71,6 +71,17 @@
...
 #     4    print "success!"
 #
+# For php, scripts the second line can't start with #!, as # is not
+# the comment character in php (even though php ignores #! on the
+# *first* line of a script).  So, instrument a php script like this,
+# using -- instead of # on the second line:
+#
+#     1    #!/bin/bash /path/to/sbang
+#     2    <?php  #/long/path/to/lua with arguments
+#     3    >?
```
Comment (adamjstewart) [inline bin/sbang]: This line needs to be updated too.

Comment (tgamblin) [reply bin/sbang]: fixed

**Review by adamjstewart — COMMENTED** [review]
This PR only works for Spack-installed PHP, right? It won't work if PHP is externally installed and other packages depend on it?

**Comment by tgamblin** [comment]
I'm not 100% sure.  The patches here modify `phar`, which is a thing that's used to create `php` packages.  It seems like @tkameyama thinks that `phar` itself has to patch those (as `phar` archives apparently contain UTF-8, and contain binary stuff, and expect to set checksums based on the contents of the package, so we can't patch them after the fact).

I don't see why normal `php` scripts that don't use `phar`  wouldn't be patched by Spack already (that's been added to the hook).  Question is what do we do with `phar` files created by a non-patched `phar`.  I dunno enough about `php` to say how that should work.

**Comment by tkameyama** [comment]
> Can you review my changes? If this is ok with you we can merge it.

I checked it.
It's OK.

**Comment by tkameyama** [comment]
> 
> I'm not 100% sure. The patches here modify `phar`, which is a thing that's used to create `php` packages. It seems like @tkameyama thinks that `phar` itself has to patch those (as `phar` archives apparently contain UTF-8, and contain binary stuff, and expect to set checksums based on the contents of the package, so we can't patch them after the fact).

phar is a special casse.
Normal php script is applied Normal sbang rule.
If  the PHP script's #! line is too long, the script will be replace to use sbang.

**Pull request merged**